read-fonts/src/tables/truetype/bytecode/decode.rs

```diff
@@ -0,0 +1,237 @@
+//! TrueType bytecode decoder.
+
+use super::{InlineOperands, Instruction, Opcode};
+
+/// An error returned by [`Decoder::decode`] if the end of the bytecode
+/// stream is reached unexpectedly.
+#[derive(Copy, Clone, Debug)]
+pub struct DecodeError(());
+
+impl std::fmt::Display for DecodeError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_str("unexpected end of bytecode")
+    }
+}
+
+/// Decodes instructions from TrueType bytecode.
+#[derive(Copy, Clone)]
+pub struct Decoder<'a> {
+    /// The bytecode for the program.
+    pub bytecode: &'a [u8],
+    /// The "program counter" or current offset into the bytecode.
+    pub pc: usize,
+}
+
+impl<'a> Decoder<'a> {
+    /// Creates a new decoder for the given bytecode and program counter.
+    pub fn new(bytecode: &'a [u8], pc: usize) -> Self {
+        Self { bytecode, pc }
+    }
+
+    /// Decodes the next instruction.
+    ///
+    /// Returns `None` at the end of the bytecode stream.
+    pub fn decode(&mut self) -> Option<Result<Instruction<'a>, DecodeError>> {
+        let opcode = Opcode::from_byte(*self.bytecode.get(self.pc)?);
+        Some(self.decode_inner(opcode))
+    }
+
+    fn decode_inner(&mut self, opcode: Opcode) -> Result<Instruction<'a>, DecodeError> {
+        let mut opcode_len = opcode.len();
+        let mut count_len = 0;
+        // If the opcode length is negative the next byte contains the number
+        // of inline operands and |opcode_len| is the size of each operand.
+        // <https://gitlab.freedesktop.org/freetype/freetype/-/blob/57617782464411201ce7bbc93b086c1b4d7d84a5/src/truetype/ttinterp.c#L7046>
+        if opcode_len < 0 {
+            let inline_count = *self.bytecode.get(self.pc + 1).ok_or(DecodeError(()))?;
+            opcode_len = -opcode_len * inline_count as i32 + 2;
```

Comment:
i would be slightly less confused by `opcode_len.abs()` but nbd

Reply:
Good call! This was initially `2 - opcode_len * next_byte` which was even more confusing.